src/main/scala/rps/controller/GameController.scala

```diff
@@ -1,40 +1,39 @@
-package rps.controller
+package rps.controller.GameController
+
+import scala.concurrent.{ExecutionContext, Future}
 
 import rps.model.{GameSummary, Move, Result}
 import Move._, Result._
 import rps.model.error._
 import rps.service.{GameService}
-import rps.repository.{GameRepository}
-import GameRepository._
-
-import scala.concurrent.{ExecutionContext, Future}
+import rps.repository.{InMemoryGameRepository}
 
 import wiro.annotation._
 
-package GameController {
-  @path("rps")
-  trait GameApi {
-    @query
-    def result(): Future[Either[NoGameInMemory, GameSummary]]
-
-    @command
-    def play(
-        userMove: Move
-    ): Future[Either[Throwable, Unit]]
-  }
-
-  class GameApiImpl(implicit ec: ExecutionContext) extends GameApi {
-    override def play(userMove: Move): Future[Either[Throwable, Unit]] =
-      Future {
-        Right(GameService.playGame(userMove))
-      }
+@path("rps")
+trait GameApi {
+  @query
+  def result(): Future[Either[NoGameInMemory, GameSummary]]
+
+  @command
+  def play(
+      userMove: Move
+  ): Future[Either[Throwable, Unit]]
+}
+
+class GameApiImpl(implicit ec: ExecutionContext, repo: InMemoryGameRepository) extends GameApi {
```

Comment:
`GameApi` should depend on `GameService`, not on the repository.

The general structure should be

`GameApi` -> `GameService` -> `GameRepository`

Also, never depend on concrete implementations. If you want to depend on the repository, depend on `GameRepository`, not on `InMemoryGameRepository`. The wiring to concrete implementations happens in `Boot`, and the cool thing is that you can change the wiring in the tests (for example) so you can pass a different implementation for - say - the repository.

Reply:
this is actually really cool...